src/meta/src/stream/scheduler.rs

```diff
@@ -199,40 +204,74 @@ impl Scheduler {
                 .insert(fragment.actors[0].actor_id, parallel_unit);
         } else {
             // Normal fragment
-            let parallel_units = if self.all_parallel_units.len() < fragment.actors.len() {
+            if self.all_parallel_units.len() < fragment.actors.len() {
                 bail!(
                     "not enough parallel units to schedule, required {} got {}",
                     fragment.actors.len(),
                     self.all_parallel_units.len(),
                 );
+            } else if fragment
+                .actors
+                .iter()
+                .any(|actor| actor.colocated_upstream_actor_id.is_some())
+            {
+                let mut parallel_unit_bitmap = HashMap::default();
+                // Record actor locations and set vnodes into the actors.
+                for actor in &mut fragment.actors {
+                    let colocated_actor_id = actor
+                        .colocated_upstream_actor_id
+                        .clone()
+                        .expect("colocated actor id must exists");
+                    assert!(!actor.upstream_actor_id.is_empty());
+                    let parallel_unit =
+                        locations.schedule_colocate_with(&[colocated_actor_id.id])?;
+                    let vnode_bitmap = locations
+                        .actor_vnode_bitmaps
+                        .get(&colocated_actor_id.id)
+                        .unwrap()
+                        .clone();
+
+                    if let Some(buffer) = vnode_bitmap.as_ref() {
+                        parallel_unit_bitmap.insert(parallel_unit.id, Bitmap::from(buffer));
+                    }
+
+                    actor.vnode_bitmap = vnode_bitmap.clone();
+                    locations
+                        .actor_locations
+                        .insert(actor.actor_id, parallel_unit);
+                    locations
+                        .actor_vnode_bitmaps
+                        .insert(actor.actor_id, vnode_bitmap);
+                }
+
+                // Construct `vnode_mapping` by merging all `vnode_bitmap` of the colocated upstream
+                // actors. Set fragment `vnode_mapping` manually.
+                fragment.vnode_mapping =
+                    Some(ParallelUnitMapping::from_bitmaps(&parallel_unit_bitmap).to_protobuf());
```

Comment:
I store `actor_vnode_bitmaps` in `ScheduledLocations` and use it to assign proper `vnode_bitmaps` for those actors having `colocated_upstream_actor_id `. Finally construct the fragment `vnode_mapping` by merging all `vnode_bitmap`s of the colocated upstream actors. cc @BugenZhao.

Reply:
LGTM for now 🥵